server/internal/worker/client.go

```diff
@@ -0,0 +1,383 @@
+// problem because we allow the getter goroutine to quit without informing the
+// sender (since it is valid for a client to close only the half of the TCP
+// connection they are using to send requests, and still expect responses to
+// come back on the other half of the connection).
+//
+// Consider the following scenario: A client connects, and sends no requests.
+// Then, it closes the connection. The getter goroutine interprets this as a
+// close of only one side of the connection, and returns. The submitter
+// goroutine is blocked reading responses from the response channel, but of
+// course no responses will ever come.
+//
+// To deal with this problem, for every connection, we spawn a background
+// goroutine that infrequently (currently once per second) polls the client to
+// see if it's still alive. If not, it destroys the whole setup (just like if a
+// user called Destroy). Since Destroy can only be called once by a user, it
+// must be safe for the underlying destroy procedure to be invoked twice - once
+// by the user, and once by this monitor goroutine.
+
+const (
+	maxOutstandingRequests = 1024
+)
+
+// A note on using Destroy:
+//
+// We need some mechanism for the code to decide that it wants to kill an
+// existing client (primarily when the user requests that the server shut down).
+// One approach is to have each ClientHandle have a global marker of some kind
+// that both the getter and the submitter check once per loop to see if they
+// should return. However, this mechanism requires that network reads and writes
+// time out quickly so that the loop is executed frequently. This in turn
+// requires intermediate partial reads or writes to be buffered and then
+// completed on the next loop. For writing, this isn't so bad - it's just a
+// matter of keeping track of how many bytes were written already. For reading,
+// it's somewhat more complicated, requiring keeping partial parse state.
+//
+// Instead, we sidestep this issue by allowing network reads and writes to block
+// as long as they want to, and instead using the Destroy method, which closes
+// the network connection. Closing a network connection causes all existing
+// pending reads and writes to immediately return, so this allows us to simply
+// call Destroy and know that outstanding calls to GetJob or SubmitResult will
+// immediately return.
+
+// A Client is a server-side handle on a connection to a client. It is capable
+// of reading job requests from the client, writing responses to the client, and
+// shutting down the connection.
+//
+// Since Clients can be backed by TCP connections, it is acceptable for GetJob
+// to return false (indicating that there are no jobs to be done - in the case
+// of a TCP connection, caused by a half-close of the connection) while
+// SubmitResult continues to return true (because writing to the connection is
+// still allowed).
+type Client interface {
```

Comment:
Because it doesn't have to be a network connection, and it particularly represents a client - not just any connection. It could be a client on the command line, for example, or something else entirely.